pkg/virt-launcher/virtwrap/cli/libvirt.go

```diff
@@ -262,11 +262,20 @@ func (l *LibvirtConnection) GetAllDomainStats(statsTypes libvirt.DomainStatsType
 	return domStats, nil
 }
 
-func (l *LibvirtConnection) GetDomainStats(statsTypes libvirt.DomainStatsTypes, flags libvirt.ConnectGetAllDomainStatsFlags) ([]*stats.DomainStats, error) {
+func (l *LibvirtConnection) GetDomainStats(statsTypes libvirt.DomainStatsTypes, flags libvirt.ConnectGetAllDomainStatsFlags) (result []*stats.DomainStats, err error) {
 	domStats, err := l.GetAllDomainStats(statsTypes, flags)
 	if err != nil {
 		return nil, err
 	}
+	// Free memory allocated for domains
+	defer func() {
+		for i := range domStats {
+			freeErr := domStats[i].Domain.Free()
+			if err == nil {
+				err = freeErr
```

Comment:
What is this? It appears prone to race conditions. if we need to pass a return value outside of a goroutine, we need to use a channel.

Reply:
There is no goroutine. It is a `defer` function call. it will be called when the whole function returns. In case the `Free()` returns an error, this deferred function passes it to the return value.

Reply:
Ok, fair point, but I'm still uncomfortable with this pattern. returning err from this cleanup-code will cause the calling function to treat the results of GetDomainStats as a fail overall.

Is it a problem if we log non-nil freeErr results and don't return an error from the function?

Reply:
Ok. I have changed it to logging.